Style: Add Deprecation sub-section (#604)
leanprover-community/leanprover-community.github.io: templates/contribute/style.md

See our [documentation requirements](doc.html) for more suggestions
and examples.

### Deprecation

Deleting, renaming, or changing theorem definitions can cause downstreams projects
that rely on these definitions to fail to compile. Any publicly exposed theorem

> definitions that are being changed should be gracefully transitioned by keeping the

Author: That's a good point. I've updated the wording to say:

> Any publicly exposed theorems and definitions that are being removed should be ...

> that rely on these definitions to fail to compile. Any publicly exposed theorem
> definitions that are being changed should be gracefully transitioned by keeping the

Reviewer: I don't see those changes...
```suggestion
that rely on these definitions to fail to compile. Any publicly exposed theorem and
definitions that are being removed should be gracefully transitioned by keeping the
```

> definitions that are being changed should be gracefully transitioned by keeping the

Author: Didn't notice that git push failed, fixed.

old theorem statement with a `@[deprecated]` attribute, to warn downstream projects to
transition away from deprecated definitions before they are deleted. Renamed theorems
can use a deprecated `alias` to the new name, while in cases where the theorem is not
being replaced, the deprecated theorem definition can be marked with an attribute, like so:

```lean4
theorem new_name : ... := ...
@[deprecated new_name (since := "YYYY-MM-DD")] alias old_name := new_name

@[deprecated "This theorem is deprecated in favor of using ... with ..." (since := "YYYY-MM-DD")]
theorem example_thm ...
```

The [`deprecate to`](/mathlib4_docs/Mathlib/Tactic/DeprecateTo.html) command can help generate
alias definitions. The `@[deprecated]` attribute should always include the deprecation date,
and either the new name for the theorem when it is being renamed or kept substantially the same,
or a string to explain how transition away from the old definition when a new version is no longer

> being provided.

Reviewer: This is maybe also a good place to point to `scripts/add_deprecations.sh`.

Author: Done.

See our [documentation requirements](doc.html) for more suggestions
and examples.

### Deprecation

Deleting, renaming, or changing declarations can cause downstreams projects
that rely on these definitions to fail to compile. Any publicly exposed theorems and
definitions that are being removed should be gracefully transitioned by keeping the

> old theorem statement with a `@[deprecated]` attribute, to warn downstream projects to

Reviewer 2: ```suggestion
old declaration with a `@[deprecated]` attribute, to warn downstream projects to

Reviewer 2: How about splitting that sentence --- something like this?
```
Any publicly exposed theorems and definitions that are being removed should be gracefully transitioned
by keeping the old declaration with a `@[deprecated]` attribute.
This warns downstream projects about the change and gives them the opportunity to adjust before the declarations are deleted.
```

transition away from deprecated definitions before they are deleted. Renamed theorems

> can use a deprecated `alias` to the new name, while in cases where the theorem is not

Reviewer 2: Renamed definitions can also use this, right? So how about writing something like
```
When a theorem or definition is renamed, please use a deprecated `alias` to the new name, like so:

> being replaced, the deprecated theorem definition can be marked with an attribute, like so:

Reviewer 2: I wonder what exactly "is not being replaced" means. What do you think about "does not have a direct replacement"? (Perhaps that's just my personal opinion, so I'd like to hear yours.)

```lean4
theorem new_name : ... := ...

> @[deprecated new_name (since := "YYYY-MM-DD")] alias old_name := new_name

Reviewer 2: Actually, isn't the syntax different? (Here's one random example: https://github.com/leanprover-community/mathlib4/pull/23841/files)
```suggestion
@[deprecated (since := "YYYY-MM-DD")] alias old_name := new_name

Author: You're right, though according to https://github.com/leanprover/lean4/pull/6112 I thought this should print a warning. It turns out that using an alias (instead of say, a def) seems to suppress the warning, though I'm not sure how.

@[deprecated "This theorem is deprecated in favor of using ... with ..." (since := "YYYY-MM-DD")]
theorem example_thm ...
```

> The `@[deprecated]` attribute should always include the deprecation date,

Reviewer 2: If I'm reading https://github.com/leanprover/lean4/pull/6112 correctly, the attribute already enforces this. So perhaps
```suggestion
The `@[deprecated]` attribute requires the deprecation date,

This warns downstream projects about the change and gives them the opportunity to adjust before the
declarations are deleted.
Renamed definitions should use a deprecated `alias` to the new name.
Otherwise, when the deprecated definition does not have a direct replacement, the definition should
be deprecated with a message, like so:

```lean4
theorem new_name : ... := ...
@[deprecated (since := "YYYY-MM-DD")] alias old_name := new_name

@[deprecated "This theorem is deprecated in favor of using ... with ..." (since := "YYYY-MM-DD")]
theorem example_thm ...
```

The `@[deprecated]` attribute requires the deprecation date, and an alias to the new declaration
or a string to explain how transition away from the old definition when a new version is no longer
being provided.

The [`deprecate to`](/mathlib4_docs/Mathlib/Tactic/DeprecateTo.html) command and
`scripts/add_deprecations.sh` script can help generate alias definitions.

Deprecations for declarations with the `to_additive` attribute should ensure the deprecation is
also tagged with `to_additive`, while for declarations generated by `simps`, this is not strictly

> required.

Reviewer 3: I think the `to_additive` one needs an example, and I'm not sure I believe the claim about `simps`.

Reviewer 2: Agreed, an example for to_additive could help.

Reviewer 4: > The sentence about simps goes back to [this message]([#mathlib4 > Deprecation policy @ 💬](https://leanprover.zulipchat.com/#narrow/channel/287929-mathlib4/topic/Deprecation.20policy/near/436744276)) by @fpvandoorn:
> 
> > I would be fine not mandating deprecations for lemmas generated by simps:
> 
> >  - They are generally not typed manually (though they could appear in simp only calls
> 
> >  - It's very hard to have a lemma generated by simps in your code without the original definition, and the simps lemma should change very predictably. (assuming the only change is renaming the def.)
> 
> 

#23959 deprecated a lot of `to_additive` lemmas, and as discussed in the comments, the process was quite a bit more involved than just adding a tag.

Author: @apnelson1 I was not aware of this, thanks for the pointer to that PR! I've removed the bit about `simps` and updated with an appropriate example.
